pkg/controller/instance/instance_controller.go

```diff
@@ -97,63 +100,71 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 				// since its linking to a bad FV.
 				return false
 			}
-			// Identify plan to be executed by this change
+
+			// Identify plan to be executed by this change.
 			var planName string
-			var ok bool
+			var planFound bool
 			if old.Spec.FrameworkVersion != new.Spec.FrameworkVersion {
-				// Its an Upgrade!
-				_, ok = fv.Spec.Plans["upgrade"]
-				if !ok {
-					_, ok = fv.Spec.Plans["update"]
-					if !ok {
-						_, ok = fv.Spec.Plans["deploy"]
-						if !ok {
-							log.Println("InstanceController: Could not find any plan to use for upgrade")
-							return false
-						}
-						ok = true // TODO: Do we need this here?
-						planName = "deploy"
-					} else {
-						planName = "update"
+				// Its an upgrade!
+				names := []string{"upgrade", "update", "deploy"}
+				for _, n := range names {
+					if _, planFound = fv.Spec.Plans[n]; planFound {
+						planName = n
+						break
 					}
-				} else {
-					planName = "upgrade"
+				}
+				if !planFound {
+					log.Printf("InstanceController: Could not find any plan to use to upgrade instance %v", new.Name)
+					return false
 				}
 			} else if !reflect.DeepEqual(old.Spec, new.Spec) {
 				for k := range parameterDifference(old.Spec.Parameters, new.Spec.Parameters) {
-					// Find the right parameter in the FV
+					// Find the spec of the updated parameter.
+					paramFound := false
 					for _, param := range fv.Spec.Parameters {
 						if param.Name == k {
-							planName = param.Trigger
-							ok = true
+							paramFound = true
+
+							if param.Trigger != "" {
+								planName = param.Trigger
+								planFound = true
+							}
+							// TODO: what should we do if more than one parameter was updated?
 						}
 					}
-					if !ok {
-						log.Printf("InstanceController: Instance %v updated parameter %v, but parameter not found in FrameworkVersion %v\n", new.Name, k, fv.Name)
-					} else if planName == "" {
-						_, ok = fv.Spec.Plans["update"]
-						if !ok {
-							_, ok = fv.Spec.Plans["deploy"]
-							if !ok {
-								log.Println("InstanceController: Could not find any plan to use for update")
-							} else {
-								planName = "deploy"
+
+					if paramFound {
+						if !planFound {
+							// The parameter doesn't have a trigger, try to find the corresponding default plan.
+							names := []string{"update", "deploy"}
+							for _, n := range names {
+								if _, planFound = fv.Spec.Plans[n]; planFound {
+									planName = n
+									planFound = true
+									break
+								}
 							}
-						} else {
-							planName = "update"
+
+							if planFound {
+								// FIXME: this can be misleading, the plan will only be used if this is the last parameter returned by `parameterDifference`.
```

Comment:
```suggestion
								// FIXME: this can be misleading, the plan will only be used if this is the first parameter returned by `parameterDifference`.
```
afterward, `planName` (on the left) or `planFound` (on the right) would be set and we won't land here.

Reply:
I think that neither my original comment nor your suggestion are completely accurate 😱.

This is all nested within the loop that starts on line 124 on the right. If for any changed param a trigger was defined (`param.trigger != ""`), then lines 142-143 will overwrite `planName` and `planFound. This will happen regardless of whether we had previously logged (and "decided") that we were going to use a default plan. This means that the plan that we logged might actually not be used, even when it corresponds to the first parameter returned by `parameterDifference`.

I think that right now we don't have a test verifying which plan will be executed if parameters with different triggers have changed. We probably also haven't documented this. We should define what we want to happen... and then clean up the code a little more to reflect that and write a test to make sure that the code is correct.

I added this FIXME here to spark this discussion... but I forgot to add a comment when posting the PR. Thanks for noticing this =).

So now my question is: which plan should we run if multiple parameters have been changed triggering different triggers? cc/ @gerred @alenkacz @runyontr

Reply:
We have another issue to discuss this one, let's not change this here. There's a lot of options and _most_ of them will require a validating admission hook.

Reply:
Cool, I wasn't aware of #354! — I'll change this comment to a TODO mentioning that we might need to revisit this part when #354 is resolved =).